### PR TITLE
Log messages to DiagnosticListener in addition to page instrumentation

### LINKED ISSUE #3281: Consider a design where BrowserLink \ page instrumentation uses EventNotification to listen to Razor page instrumentation calls.
This would simplify the design around the instrumentation we perform in RazorPage.

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #3281

### CODE REVIEW

**Comment by pranavkm** [comment]
cc @rynowak

File: test/Microsoft.AspNet.Mvc.TestDiagnosticListener.Sources/TestDiagnosticListener.cs [hunk 1]
```diff
@@ -324,5 +324,69 @@ public class OnViewComponentAfterViewExecuteEventData
...
+        {
+            public IProxyHttpContext HttpContext { get; set; }
+
+            public string Path { get; set; }
+
+            public bool IsPartial { get; set; }
+        }
+
+        public List<object> PageInstrumentationData { get; set; } =
+            new List<object>();
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.TestDiagnosticListener.Sources/TestDiagnosticListener.cs]: No need for a linebreak here

**Comment by rynowak** [comment]
:shipit:

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs [hunk 2]
```diff
@@ -1063,12 +1069,41 @@ public void EnsureRenderedBodyOrSections()
...
+                    });
+            }
         }
 
         public void EndContext()
         {
+            const string EndContextEvent = "Microsoft.AspNet.Mvc.Razor.EndInstrumentationContext";
+
             PageExecutionContext?.EndContext();
+            if (DiagnosticSource != null && DiagnosticSource.IsEnabled(EndContextEvent))
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: `DiagnosticSource?.IsEnabled(EndContextEvent) == true`

Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: Yup, I could do that now that I actually understand how `.?` works for value types - https://github.com/aspnet/HttpAbstractions/pull/461#discussion_r43961310

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs [hunk 3]
```diff
@@ -85,6 +85,12 @@ public HttpContext Context
         public IPageExecutionContext PageExecutionContext { get; set; }
 
         /// <summary>
+        /// Gets or sets a <see cref="DiagnosticSource.DiagnosticSource"/> instance used to instrument the page execution.
+        /// </summary>
+        [RazorInject]
+        public DiagnosticSource DiagnosticSource { get; set; }
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: There value in having this in DesignTime IntelliSense?

Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: DesignTime intelliense?

Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: When writing C# in your .cshtml file this'll pop up as an available member.

**Pull request closed without merging**